stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+        _STL_VERIFY(_Desired >= 0 && _Desired <= _Least_max_value,
+            "Precondition: desired >= 0, and desired <= max() (N4861 [thread.sema.cnt]/5)");
+    }
+
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(const ptrdiff_t _Update = 1) noexcept /* strengthened */ {
+        if (_Update == 0) {
+            return;
+        }
+        _STL_VERIFY(_Update > 0 && _Update <= _Least_max_value,
+            "Precondition: update >= 0, and update <= max() - counter (N4861 [thread.sema.cnt]/8)");
+
+        // We need to notify (wake) at least _Update waiting threads.
+        // Errors towards waking more cannot be always avoided, but they are performance issue.
+        // Errors towards waking less should be avoided, as they are correctness issue.
+
+        // release thread: Increment semaphore counter, then load waiting counter;
+        // acquire thread: Increment waiting counter, then load semaphore counter;
+
+        // memory_order_seq_cst for all four operations guarantees that either release thread loads
+        // the incremented value, or acquire thread loads the incremented value, or both, but not none.
+        // memory_order_seq_cst might be superfluous for some hardware mappings of C++ memory model,
+        // but from the point of C++ memory model itself it is needed, weaker orders don't work.
+
+        const ptrdiff_t _Prev = _Counter.fetch_add(static_cast<ptrdiff_t>(_Update), memory_order_seq_cst);
+        _STL_VERIFY(_Prev + _Update > 0 && _Prev + _Update <= _Least_max_value,
+            "Precondition: update <= max() - counter (N4861 [thread.sema.cnt]/8)");
+
+        const ptrdiff_t _Waiting_upper_bound = _Waiting.load(memory_order_seq_cst);
+
+        if (_Waiting_upper_bound == 0) {
+            // Definitely no one is waiting
+        } else if (_Waiting_upper_bound <= _Update) {
+            // No more waiting threads than update, can wake everyone.
+            _Counter.notify_all();
+        } else {
+            // Wake at most _Update. Though repeated notify_one() is somewhat less efficient than single notify_all(),
+            // the amount of OS calls is still the same; the benefit from trying not to wake unnecessary threads
+            // is expected to be greater than the loss on extra calls and atomic operations.
+            for (ptrdiff_t _Thd = _Update; _Thd != 0; --_Thd) {
+                _Counter.notify_one();
+            }
+        }
+    }
+
+    void _Wait(const unsigned long _Remaining_timeout) noexcept {
+        // See the comment in release()
+        _Waiting.fetch_add(1, memory_order_seq_cst);
+        ptrdiff_t _Current = _Counter.load(memory_order_seq_cst);
+        if (_Current == 0) {
+            __std_atomic_wait_direct(&_Counter, &_Current, sizeof(_Current), _Remaining_timeout);
+        }
+        _Waiting.fetch_sub(1, memory_order_relaxed);

```

Comment:
This can't be relaxed since it wants to synchronize with the load on 96; I don't think there should be any relaxed atomics in here particularly when the algorithm is relying on full seq_cst elsewhere.

Reply:
Decrement does not need to synchronize with anything, only increment needs.